### PR TITLE
Restructuring unit tests with their own sub-dirs

### PR DESCRIPTION
**Pull request by nariman87**
## Context for changes

- **Improvements**: 
    * Unit tests have been re-grouped in individual sub-dirs inside `tests/` based on error correction and software layers. This helps manage and target each test unit. 
    * `build_tests.yaml` workflow now supports executing unit tests in parallel using `pytest-xdist` package. GitHub runners have at least 2 processors, which helps speed up the pytest blocks by ~1.5 times in practice.

## Example usage and tests

- [x] Not Applicable

## Performance results justifying changes

- [x] Not Applicable

## Workflow actions and tests

See the updated [.github/workflows/build_tests.yaml](https://github.com/XanaduAI/flamingpy/blob/d3a18cd688682f6dc7408620b6a394b6fcea882f/.github/workflows/build_tests.yaml).

## Expected benefits and drawbacks

**Expected benefits:**
- unit tests now have their own sub-dirs based on error correction and software layers and are faster to run thanks to `xdist` parallelization.

## Related Github issues

- [x] Not Applicable

## Checklist and integration statements

- [x] My Python and C++ codes follow this project's coding and commenting styles as indicated by existing files. Precisely, the changes conform to given `black`, `docformatter` and `pylint` configurations. 
- [x] I have performed a self-review of these changes, checked my code (including for [codefactor](https://www.codefactor.io/repository/github/xanaduai/flamingpy/branches) compliance), and corrected misspellings to the best of my capacity. I have synced this branch with others as required.
- [x] I have added context for corresponding changes in documentation and [`README.md`](README.md) as needed.
- [x] I have added new workflow CI tests for corresponding changes, ensuring codecoverage is 95% or better, and these pass locally for me.
- [x] I have updated [`CHANGELOG.md`](.github/CHANGELOG.md) following the template. I recognize that the developers may revisit [`CHANGELOG.md`](.github/CHANGELOG.md) and the versioning, and create a Special Release including my changes.

### CODE REVIEW

File: .github/workflows/linting.yaml
Comment (sduquemesa) [inline .github/workflows/linting.yaml:42]: To recurse into subdirectories you should extend the glob pattern. Currently, if I try with the glob expression here what I get is the following
```bash
>> ls flamingpy/**.py
flamingpy/__init__.py  flamingpy/_version.py  flamingpy/simulations.py
```
Note this is shell dependent but this suggestion is the most general possible expression.

```suggestion
        run: python3 -m pylint --disable=R,W flamingpy/**/*.py tests/**/*.py
```

File: .github/pull_request_template.md
Comment (sduquemesa) [inline .github/pull_request_template.md:55]: Thank you! This is why it was rendering with spaces...

File: doc/index.rst
Comment (sduquemesa) [inline doc/index.rst:98]: This new section and the `quick_installation` section have redundant content. Perhaps we can remove the later, even more knowing that the new section looks superb! ✨

File: tests/cv/test_gkp.py
Comment (sduquemesa) [inline tests/cv/test_gkp.py:32]: Why not relying on pytest's random seeding [`pytest-randomly`](https://pypi.org/project/pytest-randomly/)? Otherwise it'd be good to log the `int_time` to be able to execute the tests with the same seed in case needed.

**Review by sduquemesa — CHANGES_REQUESTED** [review]
Thanks for this @nariman87, it does make the testing suite much more organized and focused. Overall it looks great! I have left some comments and question, in the meantime perhaps other code-owners can review the PR as well.

File: .github/workflows/linting.yaml
Comment (nariman87) [reply .github/workflows/linting.yaml:42]: Yeah, you are absolutely right, this is shell-dependent, it worked on my system, but no guarantee for it to work everywhere. I am happy with your suggestion.

File: flamingpy/utils/viz.py
Comment (nariman87) [reply flamingpy/utils/viz.py]: Actually, `too-many-arguments` appears in quite a few files and must be moved to `.pylintrc`. I will do that here and, hopefully, makes for a satisfactory resolution.

File: .github/workflows/build_tests.yaml
Comment (nariman87) [reply .github/workflows/build_tests.yaml:58]: Done.

File: tests/cv/test_gkp.py
Comment (nariman87) [reply tests/cv/test_gkp.py:85]: `sorted` ensures workers get the parameters consistently and in the right order when doing pytest parallelization. Some systems/environments may get 'different results collected for different workers' errors without it.

Interestingly, the way I set up our environments for our GitHub test workflows (with the latest pytest and proper distribution methods), we don't need `sorted` in most cases. However, if people try to run it locally they may get errors and that's why I left them there in some places.

See for example [here](https://github.com/pytest-dev/pytest-xdist/issues/432).

File: doc/index.rst
Comment (nariman87) [reply doc/index.rst:98]: Actually, the `install` is essential as it is one of the top headers consistent in Xanadu software. We may want to remove `quick_installation`? @ilan-tz, what do you think?

File: tests/cv/test_gkp.py
Comment (nariman87) [reply tests/cv/test_gkp.py:32]: Yeah, I forgot to log the integer seed. I'll do that.

File: tests/cv/test_gkp.py
Comment (sduquemesa) [reply tests/cv/test_gkp.py:86]: Oh, I see. Thanks @nariman87 for the explanation!

**Review by sduquemesa — APPROVED** [review]
Thanks @nariman87 for addressing all the comments! ✨ Let's see what are @ilan-tz's opinions on the `quick-install`/`install` issue before merging.

File: doc/index.rst
Comment (nariman87) [reply doc/index.rst:97]: Done. You may want to have another look.

File: flamingpy/cv/macro_reduce.py
Comment (ilan-tz) [inline flamingpy/cv/macro_reduce.py:22]:
```suggestion
from thewalrus.symplectic import expand, beam_splitter

```

File: tests/cv/test_macro_reduce.py
Comment (ilan-tz) [inline tests/cv/test_macro_reduce.py:41]:
```suggestion
    pad_bool = boundaries == "open"
```

File: .github/workflows/linting.yaml
Comment (ilan-tz) [inline .github/workflows/linting.yaml:42]: Does this notation include the files in flamingpy directory that are not in a subfolder?

Also the pylint test seems to pass with 10/10 -- is that suspicious?

**Review by ilan-tz — COMMENTED** [review]
Great - an important PR with several nice improvements. Just have a few small questions.

**Review by ilan-tz — APPROVED** [review]

**Pull request merged**
